packages/lesmis/src/api/queries/useMapOverlayReportData.js

```diff
@@ -0,0 +1,129 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+
+import { useQuery } from 'react-query';
+import {
+  autoAssignColors,
+  createValueMapping,
+  getSpectrumScaleValues,
+  SPECTRUM_MEASURE_TYPES,
+  getMeasureDisplayInfo,
+} from '@tupaia/ui-components/lib/map';
+import { useEntitiesData } from './useEntitiesData';
+import { get } from '../api';
+
+export const useMapOverlayReportData = ({ entityCode, mapOverlayCode, hiddenMeasures = {} }) => {
+  const { data: entitiesData, isSuccess } = useEntitiesData();
+
+  const params = {
+    shouldShowAllParentCountryResults: false,
+    type: 'mapOverlay',
+  };
+
+  const { data: measureData, ...query } = useQuery(
+    ['mapOverlay', entityCode, mapOverlayCode, hiddenMeasures, params],
+    () =>
+      get(`report/${entityCode}/${mapOverlayCode}`, {
+        params,
+      }),
+    {
+      staleTime: 60 * 60 * 1000,
+      refetchOnWindowFocus: false,
+      enabled: isSuccess && !!entitiesData,
+    },
+  );
+
+  // processMeasureInfo
+  const processedMeasureInfo = measureData ? processMeasureInfo(measureData) : null;
+
+  // processMeasureData
+  const processedMeasureData =
+    measureData && entitiesData
+      ? processMeasureData({
+          measureData: measureData.measureData,
+          entitiesData,
+          measureOptions: processedMeasureInfo.measureOptions,
+          mapOverlayCode,
+          hiddenMeasures,
+        })
+      : null;
+
+  return {
+    ...query,
+    data: { ...measureData, ...processedMeasureInfo, measureData: processedMeasureData },
+  };
+};
+
+function processMeasureData({
+  measureData,
+  entitiesData,
+  measureOptions,
+  mapOverlayCode,
+  hiddenMeasures,
+}) {
+  const measureEntities = measureData
+    .filter(measure => !hiddenMeasures.includes(measure[mapOverlayCode]))
+    .map(m => m.organisationUnitCode);
+
+  return entitiesData
+    .filter(entity => measureEntities.includes(entity.code))
+    .map(entity => {
+      const measure = measureData.find(e => e.organisationUnitCode === entity.code);
+      const displayInfo = getMeasureDisplayInfo(measure, measureOptions);
+
+      return {
+        ...entity,
+        ...measure,
+        coordinates: entity.point,
+        region: entity.region,
+        color: displayInfo.color,
+        icon: displayInfo.icon,
+        originalValue: displayInfo.originalValue,
+      };
+    });
```

Comment:
By creating an `entitiesByCode` structure, we can cut out three inner loops. Inner loops are always good things to avoid for performance reasons, but particularly when we're dealing with an array that's 10s of thousands of entities long.

I expect the same structure might be handy other places, so I recommend we create it in `useEntitiesData` so that it can return `{ entities, entityByCode, isLoading }`

```suggestion
    entitiesByCode,
  measureOptions,
  mapOverlayCode,
  hiddenValues,
}) {
  return measureData
    .filter(measure => !hiddenValues.includes(measure[mapOverlayCode]))
    .map(measure => {
      const displayInfo = getMeasureDisplayInfo(measure, measureOptions);
      const entity = entitiesByCode[measure.organisationUnitCode];

      return {
        ...entity,
        ...measure,
        coordinates: entity.point,
        region: entity.region,
        color: displayInfo.color,
        icon: displayInfo.icon,
        originalValue: displayInfo.originalValue,
      };
    });
```